### PR TITLE
Revoke leases

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

The highlight is that when a unit transition to dead, its leadership lease is revoked.
There's 4 main bits of work:

1. Update the leadership worker - we only use a raft based lease store now which auto expires leases so remove the manual expire api. Add revoke api.
2. Update the raft fsm to support revoking leases.
3. Remote legacy lease code in state and consolidate remaining legacy code needed foe upgrades.
(move code from updates.go to leaseupdates.go)
4. Update the api facades to remove the unit leadership when the unit goes to dead.

## QA steps

Deploy 3 units
watch juju status
remove the leader unit
see that juju status shows the transition to a new leader as soon as the deleted leader is gone

## Bug reference

https://bugs.launchpad.net/bugs/1469731

### CODE REVIEW

**Comment by wallyworld** [comment]
!!build!!

File: provider/dummy/environs.go [hunk 1]
```diff
@@ -417,7 +417,7 @@ func (state *environState) destroyLocked() {
 
 	if leaseManager != nil {
 		if err := worker.Stop(leaseManager); err != nil && mongoAlive() {
-			panic(err)
+			//			panic(err)
```
Comment (howbazaar) [inline provider/dummy/environs.go:420]: Put this back?

File: apiserver/common/ensuredead.go [hunk 2]
```diff
@@ -37,7 +39,13 @@ func (d *DeadEnsurer) ensureEntityDead(tag names.Tag) error {
 	if !ok {
 		return NotSupportedError(tag, "ensuring death")
 	}
-	return entity.EnsureDead()
+	if err := entity.EnsureDead(); err != nil {
+		return errors.Trace(err)
+	}
+	if d.afterDead != nil {
```
Comment (babbageclunk) [inline apiserver/common/ensuredead.go:45]: This is neat!

File: core/raftlease/fsm.go [hunk 3]
```diff
@@ -167,12 +167,7 @@ func (f *FSM) revoke(key lease.Key, holder string) *response {
...
-		for k := range entries {
-			if key != k {
-				continue
-			}
-			delete(entries, key)
-		}
+		delete(entries, key)
 		if len(entries) == 0 {
 			delete(f.groups, gKey)
 		}
```
Comment (babbageclunk) [inline core/raftlease/fsm.go]: You shouldn't need to iterate at all - the two levels of maps are effectively: 
groups[(namespace, model-uuid)] -> entries[lease] -> (holder, start, duration)

So this should work:
```
delete(entries, key)
if len(entries) == 0 {
    delete(f.groups, groupKeyFor(key))
}

**Comment by wallyworld** [comment]
$$merge$$

**Comment by manadart** [comment]
This is relevant also to https://bugs.launchpad.net/juju/+bug/1879663.

At least with this change in place, we can't now have leadership pinned to a unit that was removed - though any subsequent claimer will effectively be pinned.

I am working on a patch to ensure that removing a machine must ensure that any lease pinnings that it was responsible for are removed.

**Pull request merged**